packages/gestalt/src/Collection.js

```diff
@@ -109,7 +112,7 @@ export default class Collection extends React.PureComponent<Props, void> {
       <div className={layoutStyles.relative} style={{ width, height }}>
         {items.map(({ idx, ...style }) => (
           <div key={idx} className={layoutStyles.absolute} style={style}>
-            <Item idx={idx} />
+            <Item idx={idx} width={style.width} height={style.height} />
```

Comment:
I'm confused by this. Width and height were passed into Collection by Collage so I don't think it's necessary to pass it back to Collage's Item.

Reply:
Suppose each collection item is rendered with component `<MyCollectionItem>` which takes `width` and `height` as props:
```jsx
function MyCollectionItem(props: { width: number, height: number, idx: number }) {
  ...
}
```

Without this change, `<MyCollection>`, which renders a list of `<MyCollectionItem>` in `<Collection>`, will look like this:

```jsx
function MyCollection(props) {
  const layout = computeLayout(props);
  return (
    <Collection
     layout={layout}
      Item={({ idx }) => (
        <MyCollectionItem
          width={layout[idx].width}
          height={layout[idx].height}
          idx={idx}
        />
      )}
    />
  );
}
```
where the anonymous function effectively creates a new class on every  re-render, causing the unmount and re-mount.

It's possible to avoid this problem by memoizing layout and pass a instance method:
```jsx
class MyCollection extends React.PureComponent {
  layout = computeLayout(this.props);

  renderItem = () => (
    <MyCollectionItem
      width={this.layout[idx].width}
      height={this.layout[idx].height}
      idx={idx}
    />
  );

  render() {
    return <Collection layout={this.layout} Item={this.renderItem} />;
  }
}
```
But it requires keeping `layout` updated when props are updated and doesn't feel as clean

With this change, using a collection without worrying about its items being unmounted/remounted is simpler:
```jsx
function MyCollection(props) {
  return <Collection layout={computeLayout(props)} Item={MyCollectionItem} />;
}
```

Reply:
There's a subtle bug that you're overlooking - `Collection` is only doing a shallow comparison on the `layout` array - it's up to the caller to determine when that has changed and re-render. Only `Collage` can know the inputs to `computeLayout` and when those are invalidated.

Reply:
While doing perf profiling, I noticed that collage items are being unmounted and remounted on each re-render, and that's the problem I'm trying to solve.

I think the most elegant solution is to pass the React class component directly to the `Item` prop of Collection.
```jsx
function MyCollection(props) {
  return <Collection layout={computeLayout(props)} Item={MyCollectionItem} />;
}
```

To your point, we _can_ also turn the `<Collage>` or `<MyCollection>` into a class component:
```jsx
function getLayoutFromProps(props) {
  return ...;
}

class MyCollection extends React.PureComponent {
  renderItem = ({ idx }) => {
    const { width, height } = getLayoutFromProps(this.props)[idx];
    return (
      <MyCollectionItem
        width={width}
        height={height}
        idx={idx}
      />
    );
  };

  render() {
    const layout = getLayoutFromProps(this.props);
    return <Collection layout={layout} Item={this.renderItem} />;
  }
}
```
with some memoization of the layout computation. I'm probably missing something here, but I don't really know why this is better than passing the react component directly.